Hardware build guide (#137)
daft-engineers/vocoder: README.md

    - Debian
    - ALSA
- Hardware

>     - Raspberry Pi 3 (or newer) running a 64bit OS (we used Rasbian). For improved performance a Raspberry Pi 4 is recommended.

Reviewer: This is too long for a bullet point, we are also currently running it fine on a pi3. Would remove

Author: Moved the Pi 4 comment

Python 3 is required to automatically build and link unit tests. Python 3 comes preinstalled on most Linux distributions, and is available as a package on all others.

Although not essential, to aid performance, we recommend running the OS in the non GUI mode. This can be achieved by running:
```
sudo raspi-config
```

> Selecting 1 System Options, S5 Boot / Auto Login, B2 Console Autologin. Reboot and enjoy the resource saving.

Reviewer: You could make this visually clearer by doing a numbered list of steps, or using > to separate the steps.

Author: added some >>

## Hardware setup:
The Codec Zero should be wired up with the microphone on the left input channel and the synth on the right input channel of the auxilliary input. The Headphones or speaker shoudld be wired to either or both channels of the auxilliary output.

![Circuit diagram](./Circuit.png)

> For testing, a Sure SM58 was chosen as the microphone (using pin 2 and 3 on the XLR connector) and a software synthesizer on a pc was used as the synthesizer. A portable speaker was used as the output device. If a low sensitivity microphone is used it may need to be amplified with an external amplifier. The volume of the microphone can balanced with the synthesizer by changing the microphone gain argument when running the program.

Reviewer: ```suggestion
For testing, a Shure SM58 was chosen as the microphone (using pin 2 and 3 on the XLR connector) and a software synthesizer on a pc was used as the synthesizer. A portable speaker was used as the output device. If a low sensitivity microphone is used it may need to be amplified with an external amplifier. The volume of the microphone can balanced with the synthesizer by changing the microphone gain argument when running the program.
```

Author: I was going to ask that after you'd decided what your command line options were going to be

Reviewer: Decided it's too late to fix it, so remove that section thnx

To allow the connected devices to be changed easily, standardised connectors should be used for the inputs and outputs. Following standard convention, an XLR connector should be used for the microphone and a 1/4 TRS jack should be used for the synthesizer in and audio output.

![Hardware](./hardware.jpg)

## Configuring the Codec Zero:
The Codec zero must be set up to be used with alsa. The easiest way is to use the scripts provided on the [IQAudio github](https://github.com/iqaudio/Pi-Codec).

To find which sound card the Codec Zero is showing up as, run the following command and look for the IQAudio device, and note its card number.
```
aplay --list-devices 
```

> Using your newly found number run the following where x is the card number (probably 2 or 3)

Reviewer: ```suggestion
Using your newly found number run the following where `x` is the card number (probably 2 or 3)
```

```
git clone https://github.com/iqaudio/Pi-Codec.git
sudo alsactl restore -f Pi-Codec/IQaudIO_Codec_AUXIN_record_and_HP_playback.state x 
```

If you would like to test if the Codec Zero is working correctly, a pre recorded test sound can be played using:
```
aplay -D "hw:x,0,0" test.wav # where x is the card number
``` 

> More information, including setting up the Codec Zero including setting it as defualt device or setting it up automatically on boot available on [their website](https://www.raspberrypi.com/documentation/accessories/audio.html#codec-zero-configuration).

Reviewer: ```suggestion
For more advanced setup information (including automatic setup and default device) see the guidance on the [Raspberry Pi website.](https://www.raspberrypi.com/documentation/accessories/audio.html#codec-zero-configuration).
```

To find which sound card the Codec Zero is showing up as, run the following command and look for the IQAudio device, and note its card number.
```
aplay --list-devices 
```
Using your newly found number run the following where x is the card number (probably 2 or 3)
```
git clone https://github.com/iqaudio/Pi-Codec.git
sudo alsactl restore -f Pi-Codec/IQaudIO_Codec_AUXIN_record_and_HP_playback.state x 
```

If you would like to test if the Codec Zero is working correctly, a pre recorded test sound can be played using:
```
aplay -D "hw:x,0,0" test.wav # where x is the card number
``` 
More information, including setting up the Codec Zero including setting it as defualt device or setting it up automatically on boot available on [their website](https://www.raspberrypi.com/documentation/accessories/audio.html#codec-zero-configuration).
## Software setup:
Once you have installed the prerequisites and configured your hardware you should clone the repository and run the setup script.

```
git clone https://github.com/daft-engineers/vocoder.git
cd vocoder
chmod +x setup.sh
./setup.sh
```

The project is now ready to be built.

> <!-- need to add codec zero setup instructions  -->

Author: superfluous, removed

```
cd build
make
cd unit_tests  # run unit tests
ctest
```

## Running the Vocoder:
The vocoder can be run with: 

> <!-- Add arguments things, I mention microphone gain higher up-->

Author: We were talking about adding command line arguments for the microphone gain and the hw device id, this is a reminder that they need to go in eventually

make
cd unit_tests  # run unit tests

> ctest

Reviewer: You can actually do:

```
cd build/unit_tests
make
ctest
```
Which I think is neater. Might also be nice to tell our users about the `-j <n>` flag

Author: I think leaving out -j is fine, makes it simpler and works fine without
